michel2323/ColPack.jl
# Big refresh, slightly breaking

Hey there @michel2323!

This PR fixes #9 by modernizing and cleaning up the package, in the hope of getting it the appreciation it deserves. I tried to apply the current best practices in terms of package structure, as described in my blog https://modernjuliaworkflows.github.io/. I also added a lot of documentation, and a few breaking changes which I feel make sense (see below). However, the basic functionalities were not altered.

These changes affect ADNLPModels.jl (ping @amontoison) but the fixes are easy to apply so I'll take care of them if you accept.

**Source code**

- BREAKING: Remove `AbstractColoring`, `AbstractOrdering` and the lists `COLORINGS` and `ORDERINGS`
- BRERAKING: Replace them by single structs `ColoringMethod` and `ColoringOrder` with one string attribute (instead of an abstract type)
- BREAKING: Implement `d1_coloring` and friends as shortcuts to these structs (instead of identical but separate structs)
- BREAKING: Change the `ColPackColoring` to use references instead of length-1 vectors
- Replace `ccall` with the macro `@ccall`
- Structure into subfiles

> Strictly speaking, since nothing was documented, these changes could be considered non-breaking, because in Julia the docs _is_ the API ;)

**Documentation**

- Set up a Documenter.jl pipeline for a proper docs website
- Add tutorial page
- Add API reference page
- Write docstrings for _everything_ (please double check that I understood the code correctly)
- Clean up README

**Tests**

- Fix the source of randomness for the sparse matrix with StableRNGs.jl (otherwise it depends on the value of Julia)
- Adjust the number of colors we expect with each coloring
- Add formal test with Aqua.jl and JET.jl
- Add doctests with Documenter.jl
- Structure into subfiles
- Run Julia 1.6 in CI

**Example**

- Remove example file (see #10)

**Esthetics**

- Add `.JuliaFormatter.toml` file for uniform style enforcement, apply formatting with JuliaFormatter.jl


**Compat**

- Bump version to v0.4.0
- Add compat bounds on Julia stdlibs
- Only test with Aqua.jl and JET.jl on Julia v1.10 and above

**CI**

- Add `dependabot.yml` to track outdated dependencies
- Modernize `Test.yml` and `TagBot.yml`
- Add `Documentation.yml` to build a docs website and deploy it
